warehouse/oidc/models.py

```diff
@@ -161,28 +149,59 @@ def verify_claims(self, signed_claims: SignedClaims):
 
     @property
     def provider_name(self):  # pragma: no cover
-        # Only concrete subclasses of OIDCProvider are constructed.
+        # Only concrete subclasses are constructed.
         return NotImplemented
 
     @property
     def provider_url(self):  # pragma: no cover
-        # Only concrete subclasses of OIDCProvider are constructed.
+        # Only concrete subclasses are constructed.
         return NotImplemented
 
 
-class GitHubProvider(OIDCProvider):
-    __tablename__ = "github_oidc_providers"
-    __mapper_args__ = {"polymorphic_identity": "github_oidc_providers"}
-    __table_args__ = (
-        UniqueConstraint(
-            "repository_name",
-            "repository_owner",
-            "workflow_filename",
-            name="_github_oidc_provider_uc",
-        ),
+class OIDCProvider(OIDCProviderMixin, db.Model):
+    __tablename__ = "oidc_providers"
+
+    discriminator = Column(String)
+    projects = orm.relationship(
+        Project,
+        secondary=OIDCProviderProjectAssociation.__table__,  # type: ignore
+        backref="oidc_providers",
+    )
+    macaroons = orm.relationship(
+        Macaroon, backref="oidc_provider", cascade="all, delete-orphan", lazy=True
     )
 
-    id = Column(UUID(as_uuid=True), ForeignKey(OIDCProvider.id), primary_key=True)
+    __mapper_args__ = {
+        "polymorphic_identity": "oidc_providers",
+        "polymorphic_on": discriminator,
+    }
+
+
+class PendingOIDCProvider(OIDCProviderMixin, db.Model):
+    """
+    A "pending" OIDC provider, i.e. one that's been registered by a user
+    but doesn't correspond to an existing PyPI project yet.
+    """
+
+    __tablename__ = "pending_oidc_providers"
+
+    discriminator = Column(String)
```

Comment:
This is duplicated in both `OIDCProvider` and `PendingOIDCProvider`, should it be in the `OIDCProviderMixin`?

Reply:
Yeah, I think so! I'll do that.

Reply:
Oh, I remember why I didn't do this -- `discriminator` is needed for `polymorphic_on`, which is specialized for each separate polymorphic hierarchy.

Reply:
Doesn't it just need to be different (in the database) to support polymorphism? Why does it need to be on both classes?

Reply:
Yeah, I overthought it -- I thought for some reason this wouldn't work:

```
    __mapper_args__ = {
        "polymorphic_identity": "oidc_providers",
        "polymorphic_on": OIDCProviderMixin.discriminator,
    }
```

...but it does, so I'll make these changes 🙂

Reply:
ef67edd3